rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+| Start Date   | 08/05/2022                        |
+| Category     | enhancement,feature                 |
+| RFC PR       | https://github.com/kubewarden/rfc/pull/13  |
+| State        | **IN REVIEW**                     |
+
+
+# Summary
+
+Kubewarden will generate audit reports that operators can use to identify all
+the resources that are violating enforced/monitoring policies.
+
+The reports will be periodically refreshed by a background job.
+
+> **Note:** this RFC does not focus on how these reports are created or kept updated.
+> This is going to be covered by another RFC.
+
+# Motivation
+
+Currently Kubewarden can be used to evaluate Kubernetes resources only when a relevant
+operation happens: creation, update, deletion of the resource. However, policies evolve over time.
+More policies can be added or they can be redeployed with different settings.
+As a result of that, a Kubernetes resource that was marked as valid even a couple of hours ago,
+could be instead rejected later on because of a policy change.
+
+We want to implement a audit check that inspects all the resources that are already
+present inside of the cluster and flags all the ones that are violating at least one of
+the currently enforced policies.
+
+The results of these evaluations must be accessible by operators and end users
+in an easy way.
+
+## Examples / User Stories
+
+As a operator, I want to know if some resource in my cluster violates the latest policies.
+
+As a user, I want to know if some of resource that I own violates the cluster compliance policies.
+
+As a Kubernetes developer, I want access to the audit checks results to allow me pro-grammatically act upon it.
+
+# Detailed design
+
+Kubewarden will generate audit reports using the
+[PolicyReport](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report)
+format that is standardized by the Kubernetes SIG "wg-policy".
+
+The audit reports are defined as set of Kubernetes CRDs. The CRDs are pretty
+flexible, they can be used to store audit
+results from Kubernetes policy engines (like Kubewarden or Kyverno) but can
+also be used to record audit results from other kind of tools (like Kubernetes
+security scanning tools as Falco or Kubebench).
+
+Kubewarden will generate one `PolicyReport` object per each Namespace found inside of
+the cluster.
+Plus, it will maintain one single `ClusterPolicyReport` object to keep track
+of all the cluster-wide Kubernetes resources.
+
+## Namespace report
+
+A Namespace can be "affected" both by `ClusterAdmissionPolicy` and by `AdmissionPolicy` policies.
+When auditing a Namespace, only some Kubewarden policies are going to be interested
+about the events happening inside of that Namespace. In addition to that,
+the same Kubernetes resource could be evaluated by multiple policies. This will
+lead to several policy evaluations being performed.
+
+The `PolicyReport` is a namespaced CR that provides these information:
```

Comment:
Things that I find missing:
- How atomically is a PolicyReport updated?
  - Does the PolicyReport contain all the evaluations for the *last run* of the audit of the namespace? (if results are present in the report, it is assumed that any result is the latest)
  - Or the *last run* of the resource? (if a result for resource is present in the report, it is assumed that the results are the latest for that specific resourse, but may not be for other resources)
- How frecuently are PolicyReportResults garbage-collected?
  - Do we only save the latest Result per (resource, policy) tuple?
  - Do we save older Results? How do we configure how many? on Kubewarden, or the PolicyReport itself?
    If we save older results, I fear it will not be efficient to try to find specific results for a resource.

Reply:
we should definitely think about how to keep the PolicyReport up to date. This is a complex topic that might be worth to cover in a different RFC as pointed out [in the unresolved questions](https://github.com/kubewarden/rfc/pull/13/files#diff-8540a44feec39379c6de0278dc397e7fbd508815de206a73d8d86a12f9978138R322)

Reply:
All of that should be handed by another RFC